bbl_screen-patch/patches/printerui/qml/x1plus/GcodeGenerator.js

```diff
@@ -55,106 +13,48 @@ function createGcode(command, params = {}) {
-            break;
-    }
-    gcode += M1002({action_code:1,action:0});
-    return gcode;
-}
-function macros_vibrationCompensation(f1,f3,nozzleTemp,bedTemp) {
-    let f2 = Math.floor((f3-f1)*0.5); //midpoint
-    
-    let header = M1002({action_code:13,action:0});
-    let footer = M1002({action_code:0,action:0});
-    if (nozzleTemp>0 && nozzleTemp <= 300){
-        header += M109(nozzleTemp);
-        footer += M109(0);
-    }
-    if (bedTemp>0 && bedTemp <= 300){
-        header += M140(bedTemp,true);
-        footer += zzzM140(0,false);
-    }
 
-    let gcode = M73(0,3); // update timeline
-    gcode += M201(100); // set Z max accel
-    gcode += G90(); // absolute coords
-    gcode += M400(1); // pause
-    gcode += M17(1.2, 1.2, 0.75); // stepper current Z=0.75
-    gcode += M201(1000); // set Z max accel
-    gcode += G28(HOMING.XYZ); // home
-    gcode += G92_1(); // unknown Bambu shaper code
-    gcode += G0({x: 128, y: 128, z: 5, accel: 2400}); // move
-    gcode += header;
-    gcode += M400(1) // pause
-
-    //SWEEP
-
-    gcode += M1002({action_code:3,action:0});
-    gcode += M970({axis: 1, a: 7, f_low: f1    , f_high: f2, k: 0}); // do the sweep
-    gcode += M73(25,3);
-    gcode += M970({axis: 1, a: 7, f_low: f2 + 1, f_high: f3, k: 1});
-    gcode += M73(50,2);
-    gcode += M974(1);
-    gcode += M970({axis: 0, a: 9, f_low: f1    , f_high: f2, h: 20, k: 0});
-    gcode += M73(75,1);
-    gcode += M970({axis: 0, a: 9, f_low: f2 + 1, f_high: f3, k: 1});
-    gcode += M73(100,0);
-    gcode += M974(0);
-    gcode += M500(); // save settings
-    gcode += M975(true); // enable shaper
-    
-    gcode += footer;
-    gcode += G0({x: 65, y: 260, z: 10, accel: 1800});
 
-    gcode += M400(1);
-    return gcode;
-}
-function macros_ABL() {
-    let gcode = M1002({action_code:0,action:1});
-    gcode += M622(1);
-    gcode += M1002({action_code:1,action:0});
-    gcode += G29();
-    gcode += M400(0);
-    gcode += M500();
-    gcode += M623();
-    return gcode;
+/** 
+ * Bambu placeholders for gcode
+ * These can be used in conditional expressions with Gcode to write fairly complex macros
+ * More placeholders and info about placeholders: 
+ * https://wiki.bambulab.com/en/software/bambu-studio/placeholder-list
+*/
+const placeholders = {
+    layer_num: () => "{layer_num}",
+    layer_z: (dz) => `{layer_z + ${dz}}`,
+    total_layer_count: () => "[total_layer_count]",
+    old_retract_length: () => "[old_retract_length]",
+    new_retract_length: () => "[new_retract_length]",
+    old_filament_temp: () => "[old_filament_temp]",
+    new_filament_temp: () => "[new_filament_temp]",
+    filament_type: (idx) => `{filament_type[${idx}]}`,
+    x_after_toolchange: () => "[x_after_toolchange]",
+    y_after_toolchange: () => "[y_after_toolchange]",
+    z_after_toolchange: () => "[z_after_toolchange]",
+    old_filament_e_feedrate: () => "{old_filament_e_feedrate}",
+    new_filemnt_e_feedrate: () => "{new_filemnt_e_feedrate}",
```

Comment:
Is this our typo, or theirs?  If theirs, do we want to homogenize it here?

Reply:
This is their typo. I will fix ours property name